### PR TITLE
refactor: remove DocumentEmbeddings from in memory vector store

### PR DESCRIPTION
**Pull request by marieaurore123**
1. Refactor in memory vector store
2. Remove `VectorStore` trait
3. All other changes are side-effects of this

### CODE REVIEW

File: rig-core/examples/rag_dynamic_tools.rs [hunk 1]
```diff
@@ -163,10 +160,20 @@ async fn main() -> Result<(), anyhow::Error> {
...
+                .map(
+                    |DocumentEmbeddings {
+                         id,
+                         document,
+                         embeddings,
+                     }| { (id, document, embeddings) },
+                )
+                .collect(),
+        )?
+        .index(embedding_model);
```
Comment (cvauclair) [inline rig-core/examples/rag_dynamic_tools.rs:176]: This is not ideal, but would presumably get fixed in subsequent PR

File: rig-core/src/vector_store/in_memory_store.rs
Comment (cvauclair) [inline rig-core/src/vector_store/in_memory_store.rs:1]: In this entire file, rename type parameter `D` to `T` to follow conventions (using other type parameter names is warranted when the type parameter has strict bounds, e.g.: `M: CompletionModel`, but in this case it's generic enough that `T` is more appropriate.

File: rig-core/src/vector_store/in_memory_store.rs [hunk 2]
```diff
@@ -7,28 +7,29 @@ use std::{
...
+    /// Hashmap value is a tuple of the serializable document and its corresponding embeddings.
+    embeddings: HashMap<String, (D, Vec<Embedding>)>,
 }
 
-impl InMemoryVectorStore {
+impl<D: Serialize + Eq> InMemoryVectorStore<D> {
     /// Implement vector search on InMemoryVectorStore.
     /// To be used by implementations of top_n and top_n_ids methods on VectorStoreIndex trait for InMemoryVectorStore.
-    fn vector_search(&self, prompt_embedding: &Embedding, n: usize) -> EmbeddingRanking {
+    fn vector_search(&self, prompt_embedding: &Embedding, n: usize) -> EmbeddingRanking<D> {
```
Comment (cvauclair) [inline rig-core/src/vector_store/in_memory_store.rs:26]: Add unit tests for `vector_search` (at least one with multiple embeddings per item). 

Populate the store with fake embeddings with very few dimensions but make sure that are sensible. I.e.: Make the matching vectors close to each other (e.g.: `[0.1, 0.1, 0.5]` and `[0.1, 0.1, 0.6]`) and non-matching vectors away from each other (e.g.: `[0.1, 0.1, 0.5]` and `[-1.0, -0.6, 0.1]`)

**Review by cvauclair — CHANGES_REQUESTED** [review]

File: rig-core/src/vector_store/in_memory_store.rs [hunk 3]
```diff
@@ -56,84 +52,69 @@ impl InMemoryVectorStore {
         tracing::info!(target: "rig",
             "Selected documents: {}",
             docs.iter()
-                .map(|Reverse(RankingItem(distance, id, _, _))| format!("{} ({})", id, distance))
+                .map(|Reverse(RankingItem(distance, id, _, embed_doc))| format!("{} ({}). Specific match: {}", id, distance, embed_doc))
```
[the remark this answers is not in the document]
Comment (garance-buricatu) [reply rig-core/src/vector_store/in_memory_store.rs:55]: Ok that makes sense but i though the whole point of having the document in the Embedding object was for debugging. So maybe i'll log it with tracing::debug! ?

File: rig-core/src/vector_store/in_memory_store.rs
Comment (garance-buricatu) [reply rig-core/src/vector_store/in_memory_store.rs:1]: But T is already in use by `top_n` and `get_document`

Comment (cvauclair) [reply rig-core/src/vector_store/in_memory_store.rs:1]: Ok, I think we'll change `VectorStoreIndex` to have a trait-level generic type anyway so this won't be a problem after that

File: rig-core/src/vector_store/in_memory_store.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (cvauclair) [reply rig-core/src/vector_store/in_memory_store.rs:55]: The user can decide to log the full document after generating the embedding if they want to inspect the data, but Rig shouldnt pollute the logs with large data dumps ourselves

**Review by cvauclair — CHANGES_REQUESTED** [review]

**Pull request merged**